File: ui/components/ui/ProducerSession/producer-session/producer-session-config.ts
import * as pb from '../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { ProducerSessionTask, producerSessionTaskFromPb, producerSessionTaskToPb } from './producer-session-task/producer-session-task';

export type ProducerSessionConfig = {
  type: 'producer-session-config',
  sessionId: string,
  tasks: ProducerSessionTask[]
};

export function producerSessionConfigFromPb(v: pb.ProducerSessionConfig): ProducerSessionConfig {
  return {
    type: 'producer-session-config',
    sessionId: v.getSessionId(),
    tasks: v.getTasksList().map(producerSessionTaskFromPb)
  };
}

export function producerSessionConfigToPb(v: ProducerSessionConfig): pb.ProducerSessionConfig {
  const configPb = new pb.ProducerSessionConfig();
  configPb.setSessionId(v.sessionId);
  configPb.setTasksList(v.tasks.map(producerSessionTaskToPb));

  return configPb;
}

File: server/src/main/scala/producer/producer_session/ProducerSessionConfig.scala
package producer.producer_session

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import _root_.producer.producer_session.producer_session_task.ProducerSessionTask

case class ProducerSessionConfig(
    tasks: Vector[ProducerSessionTask]
)

object ProducerSessionConfig:
    def fromPb(v: pb.ProducerSessionConfig): ProducerSessionConfig =
        ProducerSessionConfig(
            tasks = v.tasks.map(ProducerSessionTask.fromPb).toVector
        )

    def toPb(v: ProducerSessionConfig): pb.ProducerSessionConfig =
        pb.ProducerSessionConfig(
            tasks = v.tasks.map(ProducerSessionTask.toPb)
        )
